static/dependencies.config.php

```diff
@@ -221,5 +221,10 @@
 	],
 	Network\IHTTPRequest::class => [
 		'instanceOf' => Network\HTTPRequest::class,
-	]
+	],
+	Friendica\Model\Host::class => [
+		'constructParams' => [
+			$_SERVER,
```

Comment:
What is the `$_SERVER` variable doing here?

Reply:
It is provided as the `$server` parameter of the `Model\Host` constructor by Dice when it is instantiated as part of a chained dependency. For example, calling `DI::lock()` orders Dice to create an instance of the `ILock` interface. But in PHP interface cannot be instantiated. Instead, we added a rule to `static/dependencies.config.php` to make Dice create a `LockFactory` instance. The Dice looks at the constructor of this class which is the following:
```php
public function __construct(CacheFactory $cacheFactory, IConfig $config, Database $dba, LoggerInterface $logger, Host $host)
```
Dice then proceeds to create or retrieve (if it was already created) instances of the `CacheFactory`, `IConfig`, `Database`, `Loggerinterface` and `Host` classes.

In the case of `Model\Host`, it cannot guess what it should provide for the `$server` parameter since it's an array and not a class that it could instantiate. So we have to provide the parameter value in the `static/dependencies.config.php`.

Reply:
Okay. This whole construct is still quantum physics to me - while I'm stuck a newtonian physics.

Reply:
I understand, it took me a little while to piece everything together.

Reply:
Instead of using Dice, we could write this code, starting with the "first" dependency:
```php
$cache = new Core\Config\Cache();
$profiler = new Util\Profiler($cache);
/**  @var LoggerInterface */
$logger = new Util\Logger\VoidLogger();
$dba = new Database($cache, $profiler, $logger, $_SERVER);

$configModel = new Model\Config\Config($dba);

/** @var IConfiguration */
$config = new PreloadConfig($cache, $configModel);

//
// Here's the "new" magic:
//
$host = new Host($dba, $_SERVER);

/** @var ILock */
$lock = new DatabaseLock($dba, $host->getId(), getmypid());
```
That's how "dependency injection" works in detail. It looks at the constructor of the requested class and then it recursive checks the constructor of each dependency and their constructor until it finds the very first dependency without any other dependency (in Friendica, it's the `Core\Config\Cache`) and from this point, it creates every class. This system/pattern is called "inversion of control".

Why not per hand:
- It's repetitive the same and complex (I skipped every Factory-decisions and the creation of the "real" Logger)
- Dice caches instances so in case of a re-request, so no re-creation is needed (this is done with the `shared => true` parameter in `dependencies.config.php`)

Why do we even use this pattern?
- We can mock each dependency, passed per constructor (=> make it testable)
- It shows us the "dependency tree" and errors are instant thrown if there are dependency cycles (like `Core\Config\Cache` would have had dependencies to other classes)
- Each class, using DI, can be sure that every needed pre-dependency is computed and set right => separation of concern

But, to be honest, as @annando already notices, the disadvantage of this architectural pattern is that the complexity of the overall codebase increased (because of the implicit constructor).